src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -288,13 +257,10 @@ private static void CombineSegments(ArrayBuilder<SourceText> segments, int segme
 
                     // count how many contiguous segments are reducible
                     int count = 1;
-                    for (int j = i + 1; j < segments.Count; j++)
+                    for (int j = i + 1; j < segments.Count && segments[j].Length <= segmentSize; j++)
```

Comment:
I fixed a bug that I discovered thanks to the newly added assertions for length in `StringTextWriter` and `LargeTextWriter`.